File: src/test/java/SideOutPutTest.java
import com.chengwei.entity.SensorReading;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.DataStreamSink;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.ProcessFunction;
import org.apache.flink.util.Collector;
import org.apache.flink.util.OutputTag;


public class SideOutPutTest {
    public static void main(String[] args) throws Exception {
        //需求：冰点报警-将温度值低于32F 的温度输出到 side output。
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);

        DataStream<String> stream = env.socketTextStream("localhost",7777);
        DataStream<SensorReading> dataStream = stream.map(data ->{
            String[] fields = data.split(",");
            return new SensorReading(fields[0].trim(),Long.parseLong(fields[1].trim()),Double.parseDouble(fields[2].trim()));
        });

        SingleOutputStreamOperator<SensorReading> processedStream = dataStream.process(new FreezAlert(32.0));


        processedStream.print("processed data(main stream):");
        DataStreamSink<String> freezing_alert = processedStream.getSideOutput(new OutputTag<String>("freezing alert", TypeInformation.of(String.class))).print("alert data:");

        env.execute("job");
    }
}

class FreezAlert extends ProcessFunction<SensorReading,SensorReading>{
    Double threshold;
    OutputTag<String> alertOutPut ;

    @Override
    public void open(Configuration parameters) throws Exception {
        alertOutPut = new OutputTag<String>("freezing alert",TypeInformation.of(String.class));
    }

    public FreezAlert(Double threshold) {
        this.threshold = threshold;
    }

    @Override
    public void processElement(SensorReading value, Context ctx, Collector<SensorReading> out) throws Exception {
        if(value.getTemperature() < this.threshold){
            ctx.output(alertOutPut,"freezing alert for "+value.getId()+" "+value.getTemperature());
//            ctx.output(alertOutPut,value);
        }
        else {
            out.collect(value);
        }
    }
}

File: src/main/scala/com/atguigu/apitest/SideOutPutTest.scala
package com.atguigu.apitest

import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.functions.ProcessFunction
import org.apache.flink.streaming.api.functions.timestamps.BoundedOutOfOrdernessTimestampExtractor
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.util.Collector

object SideOutPutTest {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)
    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)

    val stream = env.socketTextStream("localhost",7777)

    val dataStream = stream.map(data => {
      val dataArray = data.split(",")
      SensorReading(dataArray(0).trim, dataArray(1).trim.toLong, dataArray(2).trim.toDouble)
    })
      .assignTimestampsAndWatermarks(new BoundedOutOfOrdernessTimestampExtractor[SensorReading](Time.seconds(1)) {
        override def extractTimestamp(element: SensorReading): Long = element.timestamp * 1000
      })

    val processedStream = dataStream
      .process( new FreezingAlert())

//    dataStream.print("input data")
    processedStream.print("processed data")  //打印的是主流
    processedStream.getSideOutput(new OutputTag[String]("freezing alert")).print("alert data")

    env.execute("window test")

  }

}

class FreezingAlert() extends ProcessFunction[SensorReading,SensorReading]{ //主输出流的数据类型
  lazy val alertOutPut:OutputTag[String] = new OutputTag[String]("freezing alert")
  override def processElement(value: SensorReading, ctx: ProcessFunction[SensorReading, SensorReading]#Context, out: Collector[SensorReading]): Unit = {

    if( value.temperature < 32.0 ){
      ctx.output( alertOutPut, "freezing alert for " + value.id + value.temperature)
    } else {
      out.collect( value )
    }
  }
}